woocommerce/wc-admin
# Add Feature Flags

Closes #1345.

This PR adds feature flags to `wc-admin 🎉.

This allows us to enable/disable features depending on the enviornment. Right now, the environments this supports are `development`, `plugin`, and `core`. Think of `development`, `staging`, and `production` in Calypso for example. You can access the flags from both PHP and JS.  The approach I used on the JS side also allows Webpack to do dead code elimination [like Gutenberg](https://github.com/WordPress/gutenberg/blob/bbdd7a70389652f483a64f66b078f992fc62f917/docs/designers-developers/developers/feature-flags.md#dead-code-elimination). For more information on all this, [view the documentation that is a part of this PR](https://github.com/woocommerce/wc-admin/blob/f129156463c5359bf086c628ecca07270565cea3/docs/feature-flags.md).

I have based this approach off of a mix of the proposed Gutenberg PR (https://github.com/WordPress/gutenberg/pull/13324) and Calypso's implementation. I prefer checking against flags in the code, instead of a specific "phase" of the plugin.

Some notes/thoughts for the future:

* How should this effect documentation (if it at all)? I don't think we will be using these flags in packages (packages have versioning) so I don't think we need to do any pruning of documents just yet... but if we start documenting a feature, we might want to prune that out for "plugin documentation" or "core documentation".
* API conditional file loading. I haven't changed any of the API loading in this PR. Some endpoints are used across features. Since these endpoints are also versioned from core, it seems like we can leave loading alone too.
* Should we have two separate `plugin` environments? I imagine at some point, we may be making changes to a released WordPress.org feature plugin, and at the same time might be cutting a GitHub preview release. Should we think about adding a `beta` environment as well?
* I added a `core.json` config file, but our core merge process is TBD so this is more of a reminder.

### Screenshots

Screenshot of devdocs and activity panel disabled in the bundled `plugin`:

<img width="1305" alt="screen shot 2019-02-08 at 1 36 33 pm" src="https://user-images.githubusercontent.com/689165/52498355-9a9c2280-2ba6-11e9-802d-aab502c2a6b2.png">

### Detailed test instructions:

* Review the documentation. Does it match the usage in this PR?
* Run `npm test` and make sure all tests pass.
* Run `phpunit` and make sure all tests pass.
* Run `npm start` (close any current running build, since the commands have changed)
* Click around various parts of `wc-admin` and WooCommerce pages. Functionally, in the development environment, nothing should have changed.
* Run `npm run-script build:release` and build a `wc-admin.zip` plugin. Upload this to a different test site (wpsandbox for example). Enable wc-admin.
* Verify that the activity panel does not display.
* Verify that the devdocs does not display.
* Verify that you can see the dashboard and analytics pages.
